Mathlib/Tactic/GCongr/Core.lean

```diff
@@ -226,79 +226,91 @@ This is used as the default side-goal discharger,
 -/
 def gcongrDischarger (goal : MVarId) : MetaM Unit := Elab.Term.TermElabM.run' do
-  trace[Meta.gcongr] "Attempting to discharge side goal {goal}"
+  withTraceNode `Meta.gcongr (return m!"{·.emoji} Attempting to discharge side goal {goal}") do
   let [] ← Elab.Tactic.run goal <|
       Elab.Tactic.evalTactic (Unhygienic.run `(tactic| gcongr_discharger))
     | failure
 
 open Elab Tactic
 
-/-- See if the term is `a = b` and the goal is `a ∼ b` or `b ∼ a`, with `∼` reflexive. -/
-@[gcongr_forward] def exactRefl : ForwardExt where
-  eval h goal := do
-    let m ← mkFreshExprMVar none
-    goal.assignIfDefeq (← mkAppOptM ``Eq.subst #[h, m])
-    goal.applyRfl
-
-/-- See if the term is `a < b` and the goal is `a ≤ b`. -/
-@[gcongr_forward] def exactLeOfLt : ForwardExt where
-  eval h goal := do goal.assignIfDefeq (← mkAppM ``le_of_lt #[h])
-
-/-- See if the term is `a ∼ b` with `∼` symmetric and the goal is `b ∼ a`. -/
-@[gcongr_forward] def symmExact : ForwardExt where
-  eval h goal := do (← goal.applySymm).assignIfDefeq h
-
-@[gcongr_forward] def exact : ForwardExt where
-  eval e m := m.assignIfDefeq e
-
-/-- Attempt to resolve an (implicitly) relational goal by one of a provided list of hypotheses,
-either with such a hypothesis directly or by a limited palette of relational forward-reasoning from
-these hypotheses. -/
-def _root_.Lean.MVarId.gcongrForward (hs : Array Expr) (g : MVarId) : MetaM Unit :=
+/-- Given a term `a < b`, build a term `a ≤ b`. -/
+@[gcongr_forward] def mkLeOfLt : ForwardExt where
+  eval h := mkAppM ``le_of_lt #[h]
+
+/-- Given a term `a ∼ b` with `∼` symmetric, build a term `b ∼ a`. -/
+@[gcongr_forward] def mkSymm : ForwardExt where
+  eval := Lean.Expr.applySymm
+
+/-- Data structure storing two arrays of expressions, one assumed to comprise proofs of equalities,
+and one assumed to comprise proofs of (non-equality) relations. -/
+structure Hypotheses where
+  /-- Array of expressions, assumed to be proofs of equalities. -/
+  (equalities : Array Expr)
+  /-- Array of expressions, assumed to be proofs of relations. -/
+  (relations : Array Expr)
+
+/-- Add a hypothesis to the "state", the stored list of hypotheses.  Also perform all the forward
+reasoning on that hypothesis made available by the `gcongr_forward` mini-tactics. -/
+def addHypothesis (h : Expr) : StateRefT Hypotheses MetaM Unit := withReducibleAndInstances do
+  if !(← isProof h) then return
+  withTraceNode `Meta.gcongr (fun _ => return m!"recording hypothesis {h} : {← inferType h}") do
+  let ⟨eqs, rels⟩ ← get
+  if h.eq?.isSome then
+    let hs : Hypotheses := { equalities := eqs.push h, relations := rels }
+    set hs
+  else
+    let tacs := (forwardExt.getState (← getEnv)).2
+    let mut rels' : Array Expr := rels.push h
+    for (n, tac) in tacs do
+        try
+          let h' ←
+            withTraceNode `Meta.gcongr (return m!"{·.emoji} trying {n} on {h}") <| tac.eval h
+          rels' := rels'.push h'
```

Comment:
I suppose a forward reasoning tactic is not supposed to create an equality, right? That might be a limitation worth documenting.

(Note to self: The converse, creating a relation from an equality (by reflexivity), is handled below in `gcongrForward`.)